File: solutions/Leetcode_318/leetcode_318.cc
#include <string>
#include <unordered_map>
#include <vector>


using namespace std;

class Solution1
{
public:
    int maxProduct(vector<string> &words)
    {
        unordered_map<int, int> mask_len;
        for (string word : words)
        {
            int mask = 0;
            for (char x : word)
            {
                mask |= 1 << (x - 'a');
            }
            mask_len[mask] = max(mask_len[mask], (int) word.size());
        }
        int max_len = 0;
        for (auto x : mask_len)
            for (auto y : mask_len)
            {
                if (!(x.first & y.first))
                    max_len = max(max_len, x.second * y.second);
            }
        return max_len;
    }
};

File: solutions/Leetcode_318/leetcode_318.py
from typing import List

class Solution1:
    def max_product(self, words: List[str]) -> int:
        mask_len = {}
        for word in words:
            mask = 0
            for x in word:
                mask |= 1 << (ord(x) - 97)
            mask_len[mask] = max(mask_len.get(mask, 0), len(word))
        max_len = 0
        for i in mask_len.keys():
            for j in mask_len.keys():
                if not i & j:
                    max_len = max(max_len, mask_len[i] * mask_len[j])
        return max_len